hydra_pywr/resultsprocessor/hydra.py

```diff
@@ -0,0 +1,402 @@
+                                                                error_on_duplicate='N'))
+                self.recorder_ra_map[(resource_type, resource_id, attribute['id'])] = recorder_name
+
+        if len(resource_attributes_to_add) > 0:
+            log.info('Adding %s new resource attributes', len(resource_attributes_to_add))
+            self._add_resource_attributes(resource_attributes_to_add)
+        return self.recorder_ra_id_map
+
+    def _add_resource_attributes(self, resource_attributes_to_add):
+        from itertools import islice
+
+        def chunked_iterable(iterable, size):
+            iterator = iter(iterable)
+            while chunk := list(islice(iterator, size)):
+                yield chunk
+
+        for chunk in chunked_iterable(resource_attributes_to_add, 100):
+            returned_new_ids = self.hydra_client.add_resource_attributes(resource_attributes=chunk)
+
+            #based on the data returned by add_resource_attributes, we need to reverse the map
+            #to make lookups easier, so the key is the (resource_id, attr_id) tuple and the value is the new resource attribute ID
+            new_ids = {}
+            for ra_id, metadata in returned_new_ids.items():
+                new_ids[tuple(metadata)] = ra_id
+
+            for i, new_ra in enumerate(chunk):
+                key = (new_ra["resource_id"], new_ra["attr_id"])
+                if key not in new_ids:
+                    continue
+                new_ra['id'] = new_ids[key]
+                if new_ra['resource_type'] == 'NETWORK':
+                    # We need to set the network ID for the resource attribute
+                    new_ra['network_id'] = new_ra['resource_id']
+                if new_ra['resource_type'] == 'NODE':
+                    # We need to set the node ID for the resource attribute
+                    new_ra['node_id'] = new_ra['resource_id']
+                recorder_name = self.recorder_ra_map[
+                    (new_ra['ref_key'], new_ra.get('node_id', new_ra.get('network_id')), new_ra['attr_id'])
+                ]
+                self.recorder_ra_id_map[recorder_name] = new_ra['id']
+
+    def make_attr_unit_map(self):
+        """
+            Create a mapping between an attribute ID and its unit, as defined
+            in the template
+        """
+        if self.hydra_template is None:
+            log.info("Cannot make unit map. Template is Null")
+            return
+        attr_unit_map = {}
+        for templatetype in self.hydra_template.templatetypes:
+            for typeattr in templatetype.typeattrs:
+                attr_unit_map[typeattr.attr_id] = typeattr.unit_id
+
+        return attr_unit_map
+
+    def make_attr_name_map(self):
+        """
+            Create a mapping between an attribute's name and itself, as defined
+            in the template
+        """
+        attr_name_map = {}
+        for templatetype in self.hydra_template.templatetypes:
```

Comment:
`make_attr_unit_map` checks for an existing template but `make_attr_name_map` doesn't.